File: Sequence Valley/Prefix Sum/Q2. Make Sum Divisible by P/Q2_MakeSumDivisiblebyP.java
import java.util.HashMap;

class Solution {
    public int minSubarray(int[] nums, int p) {

        long totalSum = 0;                 // store total sum of the array
        for (int num : nums) {             // loop through each number
            totalSum += num;               // add number to total sum
        }

        int rem = (int)(totalSum % p);     // remainder when total sum is divided by p
        if (rem == 0) return 0;            // if remainder is 0, no need to remove any subarray

        HashMap<Integer, Integer> prefixMod = new HashMap<>(); // map to store prefix sum mod and index
        prefixMod.put(0, -1);              // set base case for full prefix
        long prefixSum = 0;                // running prefix sum
        int minLength = nums.length;       // track minimum subarray length

        for (int i = 0; i < nums.length; ++i) { // loop through array
            prefixSum += nums[i];                  // update prefix sum
            int currentMod = (int)(prefixSum % p); // mod of current prefix sum
            int targetMod = (currentMod - rem + p) % p; // required mod to match

            if (prefixMod.containsKey(targetMod)) {     // if we saw this mod before
                minLength = Math.min(minLength, i - prefixMod.get(targetMod)); // update answer
            }

            prefixMod.put(currentMod, i); // store current mod with index
        }

        return minLength == nums.length ? -1 : minLength; // if unchanged, return -1 else min length
    }
}

File: Sequence Valley/Prefix Sum/Q2. Make Sum Divisible by P/Q2_MakeSumDivisiblebyP.py
class Solution(object):                      # Define Solution class
    def minSubarray(self, nums, p):           # Function to find minimum subarray
        """
        :type nums: List[int]                 # nums is a list of integers
        :type p: int                          # p is the given divisor
        :rtype: int                           # function returns an integer
        """
        totalSum = sum(nums)                  # Calculate sum of all elements
        rem = totalSum % p                    # Find remainder when divided by p

        if rem == 0:                          # If sum already divisible by p
            return 0                          # No need to remove any subarray

        prefixMod = {0: -1}                   # Store prefix mod value with index
        prefixSum = 0                         # Initialize prefix sum
        minLength = len(nums)                 # Initialize minimum length

        for i, num in enumerate(nums):        # Loop through array with index
            prefixSum += num                  # Add current number to prefix sum
            currentMod = prefixSum % p        # Get current prefix modulo p
            targetMod = (currentMod - rem + p) % p  # Required modulo to remove

            if targetMod in prefixMod:        # If required modulo exists
                minLength = min(minLength, i - prefixMod[targetMod])  
                                              # Update minimum subarray length

            prefixMod[currentMod] = i         # Store current modulo with index

        return minLength if minLength < len(nums) else -1  
                                              # Return result or -1 if not possible
